### PR TITLE
Wallet Primitive Types

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#20

# Overview

- [ ] I have added a `Cardano.Wallet.Primitive` containing primitive wallet types and operations
- [ ] I have translated properties from Lemma 2.1 in the wallet spec into property tests

# Comments

### CODE REVIEW

File: src/Cardano/Wallet/Primitive.hs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+
+-- * Block
+
+data Block = Block
+    { header
+        :: !BlockHeader
+    , transactions
+        :: !(Set Tx)
+    } deriving (Show, Generic)
```
Comment (rvl) [inline src/Cardano/Wallet/Primitive.hs:71]: `Eq` would be handy for testing.

File: src/Cardano/Wallet/Primitive.hs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        :: ![TxIn]
+        -- ^ Order of inputs matters in the transaction representation. The
+        -- transaction id is computed from the binary representation of a tx,
+        -- for which inputs are serialized in a specific order.
+    , outputs
+        :: ![TxOut]
+        -- ^ Order of outputs matter in the transaction representations. Outputs
+        -- are used as inputs for next transactions which refer to them using
+        -- their indexes. It matters also for serialization.
+    } deriving (Show, Generic)
```
Comment (rvl) [inline src/Cardano/Wallet/Primitive.hs:101]: `Ord` would be needed to insert transactions into a `Set`.

**Review by rvl — COMMENTED** [review]
Looks good so far.

File: src/Cardano/Wallet/Primitive.hs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+
+{-
+    This module contains the core primitive of a Wallet. This is roughly a
+    Haskell translation of the 'Formal Specification for a Cardano Wallet'.
+
+    It doesn't contain any particular business-logic code, but define a few
+    primitive operations on Wallet core types as well.
+-}
```
Comment (Anviking) [inline src/Cardano/Wallet/Primitive.hs:13]: ✨

File: test/unit/Cardano/Wallet/PrimitiveSpec.hs [hunk 4]
```diff
@@ -0,0 +1,179 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Wallet.PrimitiveSpec
+    ( spec
```
Comment (Anviking) [inline test/unit/Cardano/Wallet/PrimitiveSpec.hs:6]: Could we export everything implicitly, so that we can run properties in `ghci` like `quickCheck $ prop_2_1_1`?

File: test/unit/Cardano/Wallet/PrimitiveSpec.hs [hunk 5]
```diff
@@ -0,0 +1,179 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```
Comment (Anviking) [inline test/unit/Cardano/Wallet/PrimitiveSpec.hs:3]: Is this "for now"/"for this" or are you suggesting most Test modules define their own `Arbitrary` instances?

File: src/Cardano/Wallet/Primitive.hs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    deriving newtype (Semigroup, Monoid)
+
+instance NFData UTxO
+
+instance Dom UTxO where
+    type DomElem UTxO = TxIn
+    dom (UTxO utxo) = Map.keysSet utxo
+
+-- ins⋪ u
+excluding :: UTxO -> Set TxIn ->  UTxO
```
Comment (Anviking) [inline src/Cardano/Wallet/Primitive.hs:180]: Shoot me, but have we considered using the unicode symbols for matching the spec?

File: src/Cardano/Wallet/Primitive.hs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+newtype Coin = Coin
+    { getCoin :: Word64
+    } deriving stock (Show, Ord, Eq, Generic)
+      deriving newtype (Enum, Num, Real, Integral)
+
+instance NFData Coin
+
+instance Bounded Coin where
+    minBound = Coin 0
+    maxBound = Coin 45000000000000000
```
Comment (Anviking) [inline src/Cardano/Wallet/Primitive.hs:147]: If this is in `Lovelace` it would be nice to mention. Either here or for the `Coin` type.

File: test/unit/Cardano/Wallet/PrimitiveSpec.hs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        it "2.1.8) (dom u ⋃ ins) ⋪ (u ⋃ v) = (ins ⋃ dom u) ⋪ v" $ checkCoverage prop_2_1_8
+        it "2.1.9) ins⋪ u = (dom u \\ ins)⊲ u" (checkCoverage prop_2_1_9)
+
+{-------------------------------------------------------------------------------
+       Wallet Specification - Lemma 2.1 - Properties of UTxO operations
+-------------------------------------------------------------------------------}
+
+prop_2_1_1 :: (Set TxIn, UTxO) -> Property
+prop_2_1_1 (ins, u) =
+    cover 50 cond "dom u ⋂ ins ≠ ∅" (property prop)
```
Comment (Anviking) [inline test/unit/Cardano/Wallet/PrimitiveSpec.hs:64]: Properties look really nice to me!
_At first_ I found it a bit hard to parse with the important logic in the `prop` `where` clause.

I see that it's inconvenient to factor out the condition from each property though.

Mayyyybe:

```haskell
coverIntersection :: Ord a => String -> Set a -> Set a -> Bool -> Property
coverIntersection label a b prop =
    cover 50 (not $ Set.null $ a `Set.intersection` b)
    label (property prop)

prop_2_1_1 :: (Set TxIn, UTxO) -> Property
prop_2_1_1 (ins, u) = coverIntersection "dom u ⋂ ins ≠ ∅" (dom u) ins $
    (u `restrictedBy` ins) `isSubsetOf` u
```

although a bit iffy, so prob not

File: test/unit/Cardano/Wallet/PrimitiveSpec.hs [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+prop_2_1_4 (ins, u, v) =
+    cover 50 cond "(dom u ⋃ dom v) ⋂ ins ≠ ∅" (property prop)
+  where
+    cond = not $ Set.null $ Set.union (dom u) (dom v) `Set.intersection` ins
+    prop =
+        ((u <> v) `restrictedBy` ins)
+            ===
+        (u `restrictedBy` ins) <> (v `restrictedBy` ins)
+
+prop_2_1_5 :: (Set TxIn, UTxO, UTxO) -> Property
```
Comment (Anviking) [inline test/unit/Cardano/Wallet/PrimitiveSpec.hs:93]: Curios: why a tuple and not two args? Just style?

**Review by Anviking — APPROVED** [review]
Nice, lgtm! Two minor things (see comments)
- Mention that `Coin` max bound is in `Lovelace`?
- Implicit exports in spec for ghci?

then a few vaguer thoughts and questions (feel free to brush them off)

File: src/Cardano/Wallet/Primitive.hs (hunk 3 above)
Comment (KtorZ) [reply src/Cardano/Wallet/Primitive.hs:13]: I knew you would like it :wink: 
That's actually a good habit we should take. Thanks for pointing that out. Maybe it should be part of our coding standard actually :thinking: ?

File: test/unit/Cardano/Wallet/PrimitiveSpec.hs (hunk 4 above)
Comment (KtorZ) [reply test/unit/Cardano/Wallet/PrimitiveSpec.hs:6]: Not implicitly, and I believe weeder would complain about this being exported and not ran by default :/

Comment (KtorZ) [reply test/unit/Cardano/Wallet/PrimitiveSpec.hs:6]: Also, we can probably just rely on hspec for that, giving some `--match` that filter on the property title.

Comment (Anviking) [reply test/unit/Cardano/Wallet/PrimitiveSpec.hs:6]: Sure, if one for some reason would want to access the properties directly in ghci, it would still be easy to temporarily remove the export list.

File: test/unit/Cardano/Wallet/PrimitiveSpec.hs (hunk 5 above)
Comment (KtorZ) [reply test/unit/Cardano/Wallet/PrimitiveSpec.hs:3]: Pretty-much yes. Ideally, we probably want something a bit more like Hedgehog were we define generators explicitly, and not via type-class. Because generators are kinda tailored to one or more tests. For instance, for those primitive, we do not care much about the shape of an address or a block header hash, that wouldn't be true of course if we were to test how addresses are serialized / unserialized!

File: src/Cardano/Wallet/Primitive.hs (hunk 6 above)
Comment (KtorZ) [reply src/Cardano/Wallet/Primitive.hs:180]: Haha. I thought of it yes. And, that's code we will probably not use much so, that could be "nice and pretty". In practice, we probably don't want that because it's a pain in the ass to write such symbols :joy: ...

File: src/Cardano/Wallet/Primitive.hs (hunk 7 above)
Comment (KtorZ) [reply src/Cardano/Wallet/Primitive.hs:147]: :100:

File: test/unit/Cardano/Wallet/PrimitiveSpec.hs (hunk 8 above)
Comment (KtorZ) [reply test/unit/Cardano/Wallet/PrimitiveSpec.hs:64]: Though the condition and coverage threshold isn't the same, albeit identical for several. We could go and have a couple of general helpers but I think it would just mess up readability in the end, and props are already rather self contained :/ ...

File: test/unit/Cardano/Wallet/PrimitiveSpec.hs (hunk 9 above)
Comment (KtorZ) [reply test/unit/Cardano/Wallet/PrimitiveSpec.hs:93]: No. QuickCheck is better at shrinking and generating arbitrary tuples than it is at generating multiple args. Hence in general, you probably always want to have a single arg that is a tuple.

File: src/Cardano/Wallet/Primitive.hs (hunk 3 above)
Comment (Anviking) [reply src/Cardano/Wallet/Primitive.hs:13]: > Maybe it should be part of our coding standard actually

I'd like that! 😍

Comment (KtorZ) [reply src/Cardano/Wallet/Primitive.hs:13]: Then make a proposal :no_mouth:

Comment (Anviking) [reply src/Cardano/Wallet/Primitive.hs:13]: I thought you might say that 👍
